src/lib/components/Canvas.svelte

```diff
@@ -0,0 +1,157 @@
+<script context="module" lang="ts">
+  import RenderManager from '../util/renderManager';
+  import { getContext as getCTX } from 'svelte';
+
+  export const KEY = Symbol();
+
+  interface TypedContext {
+    register: RenderManager['register'];
+    unregister: RenderManager['unregister'];
+    redraw: RenderManager['redraw'];
+  }
+
+  export const getTypedContext = (): TypedContext => getCTX(KEY);
+</script>
+
+<script lang="ts">
+  import { onMount, onDestroy, setContext } from 'svelte';
+
+  export let width = 640,
+    height = 640,
+    pixelRatio: number | null = null,
+    style = '',
+    autoclear = true;
+
+  /** Class field. Only works for global classes. */
+  let clazz = '';
+
+  export { clazz as class, redraw, getCanvas, getContext };
+
+  let canvas: HTMLCanvasElement;
+  let context: CanvasRenderingContext2D | null = null;
+  let animationLoop: number;
+  let layerRef: HTMLDivElement;
+  let layerObserver: MutationObserver;
+
+  const manager = new RenderManager();
+
+  function redraw() {
+    manager.redraw();
+  }
+
+  function getCanvas() {
+    return canvas;
+  }
+
+  function getContext() {
+    return context;
+  }
+
+  if (pixelRatio === undefined) {
+    pixelRatio = 2;
+  }
+
+  function draw() {
+    if (!pixelRatio) return;
+    if (!context) return;
+    manager.render({ context, width, height, pixelRatio, autoclear });
+    animationLoop = window.requestAnimationFrame(draw);
+  }
+
+  setContext(KEY, {
+    register: manager.register,
+    unregister: manager.unregister,
+    redraw: manager.redraw
+  });
+
+  onMount(() => {
+    pixelRatio = window.devicePixelRatio;
+    const browserContext = canvas.getContext('2d');
+    if (!browserContext) {
+      throw Error("No 2D browser context found. This shouldn't happen.");
+    }
+    context = browserContext;
+
+    layerObserver = new MutationObserver(getLayerSequence);
+    layerObserver.observe(layerRef, { childList: true });
+    getLayerSequence();
+    draw();
+
+    function getLayerSequence() {
+      const sequence = [...layerRef.children]
+        // use flatmap as a casting system
+        .flatMap((child) => (child instanceof HTMLElement ? child : []))
+        .filter((layer) => layer.dataset.layerId !== undefined)
+        .map((layer) => +(layer.dataset.layerId ?? 0));
```

Comment:
Can you walk me through this change?

Reply:
`.children` returns a list of Nodes, not HTMLElements. I *would* have casted this, but given that this is external input, I wanted to avoid any weird errors caused by casting.

Thus, I use `flatMap` as a combination of `filter` and `map` (as `[]` is simply erased).

(Though rereading this now, it would be cleaner to just use `filter` and `map`, wouldn't it?)

Reply:
Yeah, I don't think the `flatMap` is necessary. Could just do `.filter((layer) => child instanceof HTMLElement && layer.dataset.layerId !== undefined)`, right?

Reply:
or even just `layer => layer.dataset?.layerId !== undefined`

Reply:
I think this may just be a case of some of TypeScript's early-designed poor methods (similar to how Object.keys and the likes aren't typed) as trying the first one still doesn't pass the type data along.

For the second one, it would require a cast to `any` first, however:

```ts
const sequence = [...layerRef.children].map((layer) => {
  if (layer instanceof HTMLElement)
    return parseInt(layer.dataset.layerId ?? "0");
  else return 0;
});
```

^ This inlines all of the other array operations into one function. Would this be a better alternative?

Reply:
Let's make the fallback return value -1 — the layer ID starts at 0, so we don't want the first renderer to be called again. You added optional chaining [here](https://github.com/LeoDog896/svelte-canvas/blob/1263794d0ba6b7762851e3e6ee4a10a6703ed67b/src/lib/util/renderManager.ts#L92) so a nonexistent ID won't be an issue.

Reply:
Should also be `else return -1`